docs/design-documents/features/storage/SecureStore/SecureStore_design.md

```diff
@@ -0,0 +1,444 @@
+	- If encrypt flag set, decrypt `dest_buf` (in place)  using `_enc_handle` handle
+	- Decrement `data_size` by `chunk_size`
+- Call `_underlying_kv` `get` API with on a local `read_cmac` variable, size 16
+- Generate CMAC on local `cmac` variable 
+- If `read_cmac` doesn't match `cmac`, return "data corrupt error"
+- Call `_rbp_kv` `get` API on a local `read_cmac` variable, key `unique_key_id` (translated to string), size 16
+- If `read_cmac` doesn't match `cmac`, return "RBP authentication error"
+- Deinitialize and free `auth_handle` and `enc_handle`
+- Release `_mutex`
+- Return OK   
+
+**get_info function**
+
+Header:  
+`virtual int get_info(const char *key, info_t *info);`
+
+Pseudo code:  
+- if not `_is_initialized` return error
+- Call `_underlying_kv` `get` API with `metadata` size and `key`
+- Fill fields in `info` according to `metadata`
+- Return OK   
+
+**remove function**
+
+Header:  
+`virtual int remove(const char *key);`
+
+Pseudo code:  
+- if not `_is_initialized` return error
+- Take `_mutex` 
+- Call `_underlying_kv` `get` API with `metadata` size and `key`
+- If not found, return "Not found" error
+- If write once flag set, return "Already exists" error
+- Call `_underlying_kv` `remove` API with `key`
+- If rollback protect flag set, Call `_rbp_kv` `remove` API with unique key ID
+- Return OK
+
+### Incremental set APIs
+
+**set_start function**
+
+Header:  
+`virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);`
+
+Pseudo code:
+- Take `_mutex`  
+- Allocate an `inc_set_handle_t` and assign in handle
+- if flags include write once flag
+	- Call `_underlying_kv` `get_info` API
+	- If key exists, return "already exists" error
+- if flags include rollback protection flag
+	- Call `_underlying_kv` `get` API with `metadata` size into `metadata` field (to fill `unique_key_id` field)
+	- If key doesn't exist, incremet `_unique_key_id_counter` and fill it in `metadata` field
+- If encrypt flag set
+	- Derive a key from device key and salt "ENC" + `unique_key_id` field
+	- Using TLS entropy function on `_entropy` handle, randomly generate `nonce` field
+	- Allocate and initialize `enc_handle` AES-CTR handle field with derived key and `nonce` field 
+- Fill all available fields in `metadata` 
+- If authentication flag set
+	- Derive a key from device key and salt "AUTH" + `unique_key_id` field
+	- Allocate and initialize `auth_handle` CMAC calculation handle field with derived key 
+	- Using `auth_handle` handle, calculate CMAC on `key` & `metadata`
+- Call `_underlying_kv` `set_start` API 	    
+- Call `_underlying_kv` `set_add_data` API with `metadata` field 	    
+- Return OK   
+
+**set_add_data function**
+
+Header:  
+`virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);`
+
+Pseudo code:
+
+- If `offset` + `data_size` > data size in handle, return error
```

Comment:
You did not mention offset before. Do you mean actual data exceeds final_data_size passed in set_start?

Reply:
Offset is stored in the handle (maybe I should be more elaborate), so yes - this checks that I don't exceed final_data_size.